File: integrations/hugo-ariada/scripts/validate-screenshot.mjs
#!/usr/bin/env node
// SPDX-FileCopyrightText: 2026 Agonist Development AB
// SPDX-License-Identifier: EUPL-1.2

import { existsSync, readFileSync } from 'node:fs';
import { inflateSync } from 'node:zlib';

const pngSignature = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);

function inspectPng(path) {
  if (!existsSync(path)) throw new Error(`Missing screenshot: ${path}`);
  const buffer = readFileSync(path);
  if (!buffer.subarray(0, 8).equals(pngSignature)) throw new Error(`Not a PNG: ${path}`);

  let offset = 8;
  let width = 0;
  let height = 0;
  let bitDepth = 0;
  let colorType = 0;
  const idat = [];

  while (offset < buffer.length) {
    const length = buffer.readUInt32BE(offset);
    const type = buffer.subarray(offset + 4, offset + 8).toString('ascii');
    const data = buffer.subarray(offset + 8, offset + 8 + length);
    if (type === 'IHDR') {
      width = data.readUInt32BE(0);
      height = data.readUInt32BE(4);
      bitDepth = data[8];
      colorType = data[9];
    } else if (type === 'IDAT') {
      idat.push(data);
    } else if (type === 'IEND') {
      break;
    }
    offset += length + 12;
  }

  if (bitDepth !== 8 || ![2, 6].includes(colorType)) {
    throw new Error(`Unsupported PNG format for validation: bitDepth=${bitDepth} colorType=${colorType}`);
  }

  const channels = colorType === 6 ? 4 : 3;
  const stride = width * channels;
  const raw = inflateSync(Buffer.concat(idat));
  let source = 0;
  let previous = Buffer.alloc(stride);
  let nonBlank = 0;
  const colors = new Set();

  for (let y = 0; y < height; y += 1) {
    const filter = raw[source];
    source += 1;
    const row = Buffer.from(raw.subarray(source, source + stride));
    source += stride;
    unfilter(row, previous, channels, filter);
    for (let x = 0; x < width; x += 1) {
      const index = x * channels;
      const red = row[index];
      const green = row[index + 1];
      const blue = row[index + 2];
      const alpha = channels === 4 ? row[index + 3] : 255;
      if (alpha > 0 && !(red > 248 && green > 248 && blue > 248)) nonBlank += 1;
      if (colors.size < 256) colors.add(`${red},${green},${blue},${alpha}`);
    }
    previous = row;
  }

  return { path, width, height, nonBlank, colors: colors.size };
}

function unfilter(row, previous, channels, filter) {
  for (let index = 0; index < row.length; index += 1) {
    const left = index >= channels ? row[index - channels] : 0;
    const up = previous[index] ?? 0;
    const upLeft = index >= channels ? previous[index - channels] ?? 0 : 0;
    if (filter === 1) row[index] = (row[index] + left) & 0xff;
    else if (filter === 2) row[index] = (row[index] + up) & 0xff;
    else if (filter === 3) row[index] = (row[index] + Math.floor((left + up) / 2)) & 0xff;
    else if (filter === 4) row[index] = (row[index] + paeth(left, up, upLeft)) & 0xff;
    else if (filter !== 0) throw new Error(`Unknown PNG filter: ${filter}`);
  }
}

function paeth(left, up, upLeft) {
  const prediction = left + up - upLeft;
  const pa = Math.abs(prediction - left);
  const pb = Math.abs(prediction - up);
  const pc = Math.abs(prediction - upLeft);
  if (pa <= pb && pa <= pc) return left;
  return pb <= pc ? up : upLeft;
}

const paths = process.argv.slice(2);
if (paths.length === 0) {
  console.error('Usage: node scripts/validate-screenshot.mjs <png> [png...]');
  process.exit(2);
}

const results = paths.map(inspectPng);
for (const result of results) {
  console.log(JSON.stringify(result));
  if (result.width < 640 || result.height < 360) throw new Error(`Screenshot too small: ${result.path}`);
  if (result.nonBlank < result.width * result.height * 0.05) throw new Error(`Screenshot appears blank: ${result.path}`);
  if (result.colors < 8) throw new Error(`Screenshot has too little color variation: ${result.path}`);
}

File: integrations/elixir-phoenix-ariada/scripts/validate_screenshot.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 Agonist Development AB
# SPDX-License-Identifier: EUPL-1.2

import struct
import sys
import zlib


def read_png(path):
    with open(path, "rb") as handle:
        data = handle.read()
    if not data.startswith(b"\x89PNG\r\n\x1a\n"):
        raise SystemExit("not a PNG")

    pos = 8
    width = height = None
    color_type = None
    bit_depth = None
    chunks = []
    while pos < len(data):
        length = struct.unpack(">I", data[pos : pos + 4])[0]
        kind = data[pos + 4 : pos + 8]
        payload = data[pos + 8 : pos + 8 + length]
        pos += 12 + length
        if kind == b"IHDR":
            width, height, bit_depth, color_type = struct.unpack(">IIBB", payload[:10])
        elif kind == b"IDAT":
            chunks.append(payload)
    if width is None or height is None:
        raise SystemExit("missing IHDR")
    return width, height, bit_depth, color_type, zlib.decompress(b"".join(chunks))


def main():
    path = sys.argv[1]
    width, height, bit_depth, color_type, raw = read_png(path)
    if width < 900 or height < 600:
        raise SystemExit(f"screenshot too small: {width}x{height}")
    if bit_depth != 8 or color_type not in (2, 6):
        raise SystemExit(f"unsupported PNG format: bit_depth={bit_depth} color_type={color_type}")

    channels = 3 if color_type == 2 else 4
    stride = width * channels
    reconstructed = []
    previous = bytearray(stride)
    pos = 0
    for _ in range(height):
        filter_type = raw[pos]
        pos += 1
        scanline = bytearray(raw[pos : pos + stride])
        pos += stride
        for index in range(stride):
            left = scanline[index - channels] if index >= channels else 0
            up = previous[index]
            up_left = previous[index - channels] if index >= channels else 0
            if filter_type == 1:
                scanline[index] = (scanline[index] + left) & 0xFF
            elif filter_type == 2:
                scanline[index] = (scanline[index] + up) & 0xFF
            elif filter_type == 3:
                scanline[index] = (scanline[index] + ((left + up) // 2)) & 0xFF
            elif filter_type == 4:
                predictor = left + up - up_left
                distances = (abs(predictor - left), abs(predictor - up), abs(predictor - up_left))
                paeth = (left, up, up_left)[distances.index(min(distances))]
                scanline[index] = (scanline[index] + paeth) & 0xFF
            elif filter_type != 0:
                raise SystemExit(f"unsupported PNG filter type {filter_type}")
        reconstructed.append(bytes(scanline))
        previous = scanline

    sample = b"".join(reconstructed[:: max(1, height // 40)])
    unique_values = len(set(sample))
    if unique_values < 16:
        raise SystemExit(f"screenshot appears blank: only {unique_values} unique byte values")

    print(f"screenshot ok: {width}x{height}, nonblank unique byte values={unique_values}")


if __name__ == "__main__":
    main()
